File: backend/app/schemas/forecasting.py
"""Pydantic schemas for the Sales Forecasting API."""
from typing import Optional
from pydantic import BaseModel, Field


class ForecastMetrics(BaseModel):
    mae: Optional[float] = None
    rmse: Optional[float] = None
    mape: Optional[float] = None
    n_test_samples: int = 0


class ModelForecastResult(BaseModel):
    model_config = {"protected_namespaces": ()}

    model_name: str
    order: list[int]
    seasonal_order: list[int]
    train_metrics: ForecastMetrics
    test_metrics: ForecastMetrics
    test_predictions: list[float] = Field(default_factory=list)
    forecast: list[float] = Field(default_factory=list)
    fitted_values: list[float] = Field(default_factory=list)


class SalesDatasetMeta(BaseModel):
    sales_dataset_id: str
    filename: str
    user_id: str
    total_rows: int
    products: list[str] = Field(default_factory=list)
    date_range_start: str
    date_range_end: str
    created_at: str


class SentimentBuildRequest(BaseModel):
    review_dataset_id: str
    product_name: Optional[str] = None


class JoinDatasetRequest(BaseModel):
    sales_dataset_id: str
    review_dataset_id: str
    product_name: str


class TrainRequest(BaseModel):
    sales_dataset_id: str
    review_dataset_id: str
    product_name: str
    order: list[int] = Field(default=[1, 1, 1])
    seasonal_order: list[int] = Field(default=[0, 0, 0, 0])
    forecast_horizon: int = Field(default=30, ge=1, le=365)
    test_fraction: float = Field(default=0.2, ge=0.05, le=0.5)


class ForecastRunSummary(BaseModel):
    run_id: str
    product_name: str
    created_at: str
    has_sentiment: bool = False
    n_train: int = 0
    n_test: int = 0
    sarima_test_metrics: Optional[ForecastMetrics] = None
    sarimax_test_metrics: Optional[ForecastMetrics] = None


class ForecastRunDetail(BaseModel):
    run_id: str
    sales_dataset_id: str
    review_dataset_id: str
    product_name: str
    order: list[int]
    seasonal_order: list[int]
    forecast_horizon: int
    test_fraction: float
    n_train: int
    n_test: int
    has_sentiment: bool
    sarima: Optional[ModelForecastResult] = None
    sarimax: Optional[ModelForecastResult] = None
    actual_dates: list[str] = Field(default_factory=list)
    actual_values: list[float] = Field(default_factory=list)
    train_dates: list[str] = Field(default_factory=list)
    test_dates: list[str] = Field(default_factory=list)
    test_actual_values: list[float] = Field(default_factory=list)
    future_dates: list[str] = Field(default_factory=list)
    created_at: str


class DailySentimentRecord(BaseModel):
    date: str
    product_name: str
    avg_score: float
    review_count: int

File: frontend/src/types/forecasting.ts
export interface ForecastMetrics {
  mae: number | null;
  rmse: number | null;
  mape: number | null;
  n_test_samples: number;
}

export interface ModelForecastResult {
  model_name: string;
  order: number[];
  seasonal_order: number[];
  train_metrics: ForecastMetrics;
  test_metrics: ForecastMetrics;
  test_predictions: number[];
  forecast: number[];
  fitted_values: number[];
}

export interface SalesDatasetMeta {
  sales_dataset_id: string;
  filename: string;
  user_id: string;
  total_rows: number;
  products: string[];
  date_range_start: string;
  date_range_end: string;
  created_at: string;
}

export interface DailySentimentRecord {
  date: string;
  product_name: string;
  avg_score: number;
  review_count: number;
}

export interface ForecastRunSummary {
  run_id: string;
  product_name: string;
  created_at: string;
  has_sentiment: boolean;
  n_train: number;
  n_test: number;
  sarima_test_metrics?: ForecastMetrics;
  sarimax_test_metrics?: ForecastMetrics;
}

export interface ForecastRunDetail {
  run_id: string;
  sales_dataset_id: string;
  review_dataset_id: string;
  product_name: string;
  order: number[];
  seasonal_order: number[];
  forecast_horizon: number;
  test_fraction: number;
  n_train: number;
  n_test: number;
  has_sentiment: boolean;
  sarima?: ModelForecastResult;
  sarimax?: ModelForecastResult;
  actual_dates: string[];
  actual_values: number[];
  train_dates: string[];
  test_dates: string[];
  test_actual_values: number[];
  future_dates: string[];
  created_at: string;
}

export interface TrainRequest {
  sales_dataset_id: string;
  review_dataset_id: string;
  product_name: string;
  order?: number[];
  seasonal_order?: number[];
  forecast_horizon?: number;
  test_fraction?: number;
}
